concent_api/api_testing_common.py

```diff
@@ -0,0 +1,132 @@
+import sys
+
+from golem_messages.exceptions      import MessageError
+from golem_messages.message import Message, ComputeTaskDef, TaskToCompute
+from golem_messages.shortcuts       import dump
+from golem_messages.shortcuts       import load
+
+import datetime
+import json
+import requests
+import http.client
+
+
+def print_golem_message(message, private_key, public_key, indent = 4):
+    assert isinstance(message, Message)
+    HEADER_FIELDS  = ['timestamp', 'encrypted', 'sig']
+    PRIVATE_FIELDS = {'_payload', '_raw'}
+    assert 'type' not in message.__slots__
+    fields = ['type'] + HEADER_FIELDS + sorted(set(message.__slots__) - set(HEADER_FIELDS) - PRIVATE_FIELDS)
+    values = [
+        type(message).__name__ if field == 'type' else
+        getattr(message, field)
+        for field in fields
+    ]
+
+    for field, value in zip(fields, values):
+        if isinstance(value, Message):
+            print_golem_message(value, private_key, public_key, indent = indent + 4)
+        else:
+            print('{}{:30} = {}'.format(' ' * indent, field, value))
+
+
+def validate_response(actual_status_code, expected_status):
+    if expected_status is not None:
+        assert expected_status == actual_status_code, f"Expected:HTTP{expected_status}, actual:HTTP{actual_status_code}"
```

Comment:
Could you replace this with a normal `if`? I don't think `assert` is what you meant here. It's for checking invariants and assuptions that are supposed to be true as long as the code is bug free. Not for validating input from user or an external server. Assertions are a debug feature, normally disabled in production.

Reply:
here `assert` is used in the same context as typically self.assertEqual from unittest is used - to check that software under test behaves as it should, that the response we have received is as inteded. Otherwise a particular test fails.

Reply:
But this is not `unittest`. Using `assert` here means that we're checking an invariant, precondition, etc. That we're stating something that universally holds unless the code has a bug.

In `unittest` we basically have no choice because of their unfortunate design choice to reuse `AssertionError` instead of creating a custom exception for reporting test failures. This conflates errors caused by failing assumptions (bugs in the testing code) with actual test failures (bugs in the code under test). I think that `unittest` was just written by someone who never seriously used `assert` for its intended purpose and did not care. We do care though.

We still keep that distinction in automated tests. You'll notice that we only ever use bare `assert` when we're making sure that all the data and environment satisfies our assumptions before we run the code under test.